File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(records):
    result = []
    id_name_match = {}
    for record in records:
        split_record = record.split()
        action = split_record[0]
        user_id = split_record[1]
        
        if action in ("Enter", "Change"):
            user_name = split_record[2]
            id_name_match[user_id] = user_name
            
        if action == "Enter":
            result.append([user_id, "IN"])
        elif action == "Leave":
            result.append([user_id, "OUT"])
            
    answer = []
    for uid, status in result:
        name = id_name_match[uid]
        if status == "IN":
            answer.append(f"{name}님이 들어왔습니다.")
        else:
            answer.append(f"{name}님이 나갔습니다.")
    return answer

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.js
function solution(records) {
    let result = [];
    let matchIdName = {};
    
    for(const record of records){
        const splitRecord = record.split(" ");
        const action = splitRecord[0];
        const userId = splitRecord[1];
        
        if(["Enter", "Change"].includes(action)){
            const userName = splitRecord[2];
            matchIdName[userId] = userName;
        }
        
        if(action === "Enter"){
            result.push([userId, "IN"]);
        } else if(action === "Leave"){
            result.push([userId, "OUT"]);
        }
    }
    
    let answer = [];
    for(const [uId, status] of result){
        const uName = matchIdName[uId];
        answer.push(status === "IN" ? `${uName}님이 들어왔습니다.` : `${uName}님이 나갔습니다.`);
    }
    return answer;
}